clients/registry-amqp/src/main/java/org/eclipse/hono/client/registry/amqp/AnnotatedCacheKey.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2021 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+package org.eclipse.hono.client.registry.amqp;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * A cache key that supports adding metadata attributes to be used to filter cache entries.
+ * <p>
+ * The metadata attributes are not taken into account by the {@link #equals(Object)} and {@link #hashCode()} methods and
+ * are therefore not part of the identity checks by the cache implementation.
+ *
+ * @param <T> The type of the cache key.
+ */
+public final class AnnotatedCacheKey<T> {
```

Comment:
this class doesn't seem to be specific to AMQP, is it? If not, then IMHO we should put it into the generic registry module or maybe even the client-common module instead.